src/databricks/labs/ucx/migration/azure_credentials.py

```diff
@@ -0,0 +1,249 @@
+import csv
+import io
+import logging
+from dataclasses import dataclass, fields
+
+from databricks.labs.blueprint.installation import Installation
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors import Unauthenticated, PermissionDenied, NotFound, InternalError
+
+from databricks.labs.ucx.assessment.azure import AzureResourcePermissions, AzureResources, \
+    StoragePermissionMapping, AzureServicePrincipalCrawler
+from databricks.labs.ucx.config import WorkspaceConfig
+from databricks.labs.ucx.framework.crawlers import StatementExecutionBackend
+from databricks.labs.ucx.hive_metastore.locations import ExternalLocations
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ServicePrincipalMigrationInfo(StoragePermissionMapping):
+    # if create access manager and managed identity for this SP
+    replace_with_access_connector: bool
+    # if a storage credential using this SP already exists
+    already_in_storage_credential: bool
+    # SP's client_secret stored in Databricks secret
+    client_secret: str
+    # if this is a managed identity
+    if_managed_identity: bool
+
+    @classmethod
+    def from_storage_permission_mapping(cls, storage_permission_mapping, **kwargs):
+        attrs = {field.name: getattr(storage_permission_mapping, field.name) for field in fields(storage_permission_mapping)}
```

Comment:
Don't use reflection! Static code analysis would miss detection bugs in your code.